src/runner_manager.py

```diff
@@ -0,0 +1,358 @@
+        )
+
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
+    def update_runner_bin(self, download_url: str) -> None:
+        """Download a runner file, replacing the current copy.
+
+        Args:
+            download_url (str): URL to download the runner binary.
+
+        TODO:
+            Deletion first cause no runner binary on download failure. See if this cause problems.
+        """
+        logger.info("Downloading runner binary from: %s", download_url)
+
+        # Remove any existing runner bin file
+        if self.runner_bin_path.exists():
+            self.runner_bin_path.unlink()
+        # Download the new file
+        response = self.session.get(download_url)
+
+        logger.info("Download of runner binary return status code: %i", response.status_code)
+
+        with self.runner_bin_path.open(mode="wb") as runner_bin_file:
+            runner_bin_file.write(response.content)
+
+        # Verify the file integrity.
+        with tarfile.open(self.runner_bin_path, "r:gz") as f:
+            try:
+                logger.debug(
+                    "Downloaded GitHub runner binary contains files: %s", ", ".join(f.getnames())
+                )
+            except tarfile.TarError:
+                logger.error("Failed to decompress downloaded GitHub runner binary.")
+                raise
+
+    def get_github_info(self) -> List[RunnerInfo]:
+        """Get information on the runners from GitHub.
+
+        Returns:
+            List of information from GitHub on runners.
+        """
+        remote_runners = self._get_runner_github_info()
+        return [
+            RunnerInfo(r.name, r.status == "online", r.status == "offline")
+            for r in remote_runners.values()
+        ]
+
+    def reconcile(self, quantity: int, resources: VirtualMachineResources) -> int:
+        """Bring runners in line with target.
+
+        Args:
+            quantity: Number of intended runners.
+            resources: Configuration of the virtual machine resources.
+
+        Returns:
+            Difference between intended runners and actual runners.
+        """
+        runners = self._get_runners()
+
+        # Clean up offline runners
+        offline_runners = [r for r in runners if not r.online]
+        if offline_runners:
+            runner_names = ", ".join(r.name for r in offline_runners)
+            logger.info("Cleaning up offline runners: %s", runner_names)
+
+            for runner in offline_runners:
+                runner.remove()
+
+        # Add/Remove runners to match the target quantity
+        online_runners = [r for r in runners if r.exist and r.online]
+        delta = quantity - len(online_runners)
+        if delta > 0:
+            logger.info("Getting registration token for GitHub runners.")
+            if isinstance(self.path, GitHubRepo):
+                token = self._github.actions.create_registration_token_for_repo(
```

Comment:
How fine grained is the creation of this token?

Reply:
The created token should expires in a hour and is used for registering runners:
https://docs.github.com/en/rest/actions/self-hosted-runners?apiVersion=2022-11-28#create-a-registration-token-for-an-organization

Reply:
And do you know if the token is the minimally possible token for the runners?

Reply:
The registration token should only have access to register runners.
There is no setting related to access scope.

This is not the personal access tokens, which can be scoped to limit access.